File: backend/app/lib/member_majors.py
"""Canonical major labels and safe abbreviation aliases.

Only include abbreviation mappings that are unambiguous in this app
(e.g. CS → Computer Science). Do not silently merge ambiguous abbreviations.
"""

from __future__ import annotations

# lowercased input → canonical display label
MAJOR_ALIASES: dict[str, str] = {
    "cs": "Computer Science",
    "c.s.": "Computer Science",
    "c.s": "Computer Science",
    "computer science": "Computer Science",
}


def normalize_major(value: str) -> str:
    """Trim, apply safe aliases, and title-case for consistent storage/display."""
    trimmed = " ".join(value.strip().split())
    if not trimmed:
        return trimmed

    lower = trimmed.lower()
    if lower in MAJOR_ALIASES:
        return MAJOR_ALIASES[lower]

    return " ".join(part.capitalize() for part in lower.split(" "))


def majors_match(left: str, right: str) -> bool:
    return normalize_major(left) == normalize_major(right)

File: frontend/src/lib/member-majors.ts
/**
 * Canonical major labels and safe abbreviation aliases.
 *
 * Only include abbreviation mappings that are unambiguous
 * (e.g. CS → Computer Science). Do not silently merge ambiguous abbreviations.
 */

const MAJOR_ALIASES: Record<string, string> = {
  cs: "Computer Science",
  "c.s.": "Computer Science",
  "c.s": "Computer Science",
  "computer science": "Computer Science",
};

export function normalizeMajor(value: string): string {
  const trimmed = value.trim().replace(/\s+/g, " ");
  if (!trimmed) {
    return trimmed;
  }

  const lower = trimmed.toLowerCase();
  if (MAJOR_ALIASES[lower]) {
    return MAJOR_ALIASES[lower];
  }

  return lower
    .split(" ")
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(" ");
}

export function uniqueNormalizedMajors(majors: string[]): string[] {
  const seen = new Set<string>();
  const result: string[] = [];

  for (const major of majors) {
    const normalized = normalizeMajor(major);
    if (!normalized || seen.has(normalized)) {
      continue;
    }
    seen.add(normalized);
    result.push(normalized);
  }

  return result.sort((a, b) => a.localeCompare(b));
}

export function memberMatchesMajors(
  member: { major?: string | null },
  selectedMajors: string[],
): boolean {
  if (selectedMajors.length === 0) {
    return true;
  }
  if (!member.major) {
    return false;
  }
  const normalized = normalizeMajor(member.major);
  return selectedMajors.some(
    (selected) => normalizeMajor(selected) === normalized,
  );
}
